internal/proxy/reverseproxy/reverseproxy.go

```diff
@@ -0,0 +1,53 @@
+package reverseproxy
+
+import (
+	"context"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// ReverseProxy is a HTTP Handler that takes an incoming request and
+// sends it to another server (jenkins), proxying the response back to
+// the client if it is received within a timeout (responseTimeout).
+// In case the handler does not recieve any response from the server
+// within the timeout, a 302 to the same URL is sent back to the client.
```

Comment:
lets start with 504 issue which openshift gateway returns when proxy fails to respond with in the timeout (30s). So the solution here is to wait until say 25s for jenkins to give jenkins-proxy a response and in case it doesn't respond by then, jenkins-proxy will send a 302 to browser. Since it is a 302 with the same url as the previous request, browser will retry the same url giving another 25s to respond.